File: concrete/src/main/java/pcb/uwu/amount/derived/area/SquareMeters.java
package pcb.uwu.amount.derived.area;

import pcb.uwu.amount.base.Candelas;
import pcb.uwu.amount.derived.optics.Illuminance;
import pcb.uwu.amount.derived.optics.Lumens;
import pcb.uwu.amount.derived.optics.Luminance;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.fundamental.AreaUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.derived.area.SquareMeterUnit.SQUAREMETER;
import static pcb.uwu.unit.derived.optics.LuxUnit.LUX;
import static pcb.uwu.unit.derived.optics.NitUnit.NIT;

public class SquareMeters extends Area {

	// region constructors

	public SquareMeters(Number value) {
		super(value, SQUAREMETER);
	}

	public SquareMeters(Number value, Magnitude magnitude) {
		super(value, magnitude, SQUAREMETER);
	}

	public SquareMeters(String value) {
		super(value, SQUAREMETER);
	}

	public SquareMeters(String value, Magnitude magnitude) {
		super(value, magnitude, SQUAREMETER);
	}

	public SquareMeters(BigDecimal value) {
		super(value, SQUAREMETER);
	}

	public SquareMeters(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, SQUAREMETER);
	}

	public SquareMeters(BigDecimalAmount amount) {
		super(amount, SQUAREMETER);
	}

	public SquareMeters(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, SQUAREMETER);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public SquareMeters plus(UnitAmount<AreaUnit> other, MathContext mathContext) {
		return new SquareMeters(plusAmount(other, mathContext));
	}

	@Override
	public SquareMeters minus(UnitAmount<AreaUnit> other, MathContext mathContext) {
		return new SquareMeters(minusAmount(other, mathContext));
	}

	@Override
	public SquareMeters multipliedBy(BigDecimal other, MathContext mathContext) {
		return new SquareMeters(multipliedByScalar(other, mathContext));
	}

	@Override
	public SquareMeters dividedBy(BigDecimal other, MathContext mathContext) {
		return new SquareMeters(dividedByScalar(other, mathContext));
	}

	// endregion

	// region composition

	public Candelas multipliedBy(Luminance luminance, MathContext mathContext) {
		return new Candelas(getAmount().multipliedBy(luminance.getAmountIn(NIT), mathContext));
	}

	public Lumens multipliedBy(Illuminance illuminance, MathContext mathContext) {
		return new Lumens(getAmount().multipliedBy(illuminance.getAmountIn(LUX), mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/derived/geometry/area/SquareMeters.kt
package pcb.uwu.amount.derived.geometry.area

import pcb.uwu.amount.base.Candelas
import pcb.uwu.amount.base.Length
import pcb.uwu.amount.base.Meters
import pcb.uwu.amount.derived.geometry.Area
import pcb.uwu.amount.derived.optics.Illuminance
import pcb.uwu.amount.derived.optics.Lumens
import pcb.uwu.amount.derived.optics.Luminance
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.MeterUnit
import pcb.uwu.unit.derived.geometry.AreaUnit
import pcb.uwu.unit.derived.geometry.SquareMeterUnit
import pcb.uwu.unit.derived.optics.LuxUnit
import pcb.uwu.unit.derived.optics.NitUnit

class SquareMeters : Area
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, SquareMeterUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, SquareMeterUnit)

    // region UnitAmount

    override fun plus(area: UnitAmount<AreaUnit>) =
        SquareMeters(this.amount + (area to SquareMeterUnit).amount)

    override fun minus(area: UnitAmount<AreaUnit>) =
        SquareMeters(this.amount - (area to SquareMeterUnit).amount)

    override fun times(scalar: Number) =
        SquareMeters(this.amount * scalar)

    override fun div(scalar: Number) =
        SquareMeters(this.amount / scalar)

    // endregion

    // region composition

    operator fun times(luminance: Luminance) =
        Candelas(amount * (luminance to NitUnit).amount)

    override fun times(illuminance: Illuminance) =
        Lumens(amount * (illuminance to LuxUnit).amount)

    operator fun div(length: Length) =
        Meters(amount / (length to MeterUnit).amount)

    // endregion

    companion object
    {
        val FOR = SquareMetersFactory()
    }
}
